src/rest/json.cpp

```diff
@@ -0,0 +1,559 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "rest/json.hpp"
+
+#include <cJSON.h>
+
+namespace otbr {
+namespace rest {
+namespace JSON {
+
+std::string String2JsonString(std::string aString)
+{
+    cJSON *     json = cJSON_CreateString(aString.c_str());
```

Comment:
```suggestion
    cJSON *     json = std::shard_ptr<cJSON>(cJSON_CreateString(aString.c_str()));
```
To avoid memory leak in any case?
Thoughts? @wgtdkp

Reply:
In cJSON, each time create a cJSON object and get its pointer, it should be delete explicitly use `cJSON_Delete` instead of delete , so if we use shared ptr, I'm not sure it is a need to explicitly assign the deconstruct callback to `cJSON_Delete`, I will investigate on it.

Reply:
@simonlingoogle  This is not easy. Because a cJSON object maybe added as a child of another cJSON object (typically, JSON map and array) and the parent object will release this child, right? @tttttangTH .

Instead of `std::shared_ptr`, I think `std::unique_ptr` fits better for our usage. And you can specify the deleter:
```c++
template<
    class T,
    class Deleter = std::default_delete<T>
> class unique_ptr;
```
https://en.cppreference.com/w/cpp/memory/unique_ptr

For the issue of adding a JSON object to another, we must call `std::unique_ptr::release` to get the raw pointer.

Reply:
But I think this is complex and not convenient, @tttttangTH will see this is not easy to use.

Reply:
Yes, when a cJSON object is added to a parent object, then delete this parent will also delete this object as well.

>  If you have added an item to an array or an object already, you mustn't delete it with cJSON_Delete. Adding it to an array or object transfers its ownership so that when that array or object is deleted, it gets deleted as well. 

So to use smart ptr, I think we need to solve the problem that how to avoid calling cJSON_Delete on one object after it is added to another object.

Reply:
> So to use smart ptr, I think we need to solve the problem that how to avoid calling cJSON_Delete on one object after it is added to another object.

`std::unique_ptr::release` is for this.

But I think it doesn't worse using smart pointers. As long as we don't expose cJSON object out of our JSON module, making sure no memory leak is not that hard.

Reply:
I think it's a bit complex to use smart pointers here and unclear if it's helpful.  
Maybe let's enhance it later.

Reply:
agree.